### PR TITLE
Arduino unit testing

### LINKED ISSUE #51: [CI] Setting-up Arduino_CI_test framework
**Description**
The Arduino_CI test .yml command lines are going to be implemented for the smart car so that all the new code snippets we are introducing to the system are implemented faultless.

**To-Do**
51.1 Arduino_CI test .yml commands should be executed for each new push.

### LINKED ISSUE #65: Arduino-Unit-Testing
Arduino unit testing shall be implemented.

### PR DESCRIPTION
**Pull request by ediz-genc**
The following methods in the sketch "smartCar.ino"  are unit tested; 

- connectHost()
- MQTTMessageInput()
  - connect()
  - subscribe()
  - onMessage()
- SR04sensorData()
- handleInput()
- emergencyBrake()
- reactToSides()
- handleSpeedTopic()
- handleAngleTopic()
- handleSpeedInput()
- handleAngleInput()
- reactToSensor()
- isClear()
- slowDownSmoothly()
- convertSpeed()

Resolves #65
Closes #65  
Resolves #51 
Closes #51 

Co-authored-by: @edizgenc2021 <gusgened@student.gu.se>

Co-authored-by: @bassamEldesouki <bassam.eldesouki.k@gmail.com>

### CODE REVIEW

File: arduino/wrappers/Utils.h [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+#else // use the syntax for C++.
+  return std::stoi(input);
+#endif
+}
+
+// In this function is tested;
+// connect() to "arduino", "public", "public",
+// subscribe() to this "/smartcar/control/#" topic and subscribe() to this "/smartcar/connectionLost" topic.
+// and onMessage() function is tested to callback().
+// handleSpeedTopic() & handleAngleTopic() are commented out because they are tested in this method.
```
Comment (iachella) [inline arduino/wrappers/Utils.h:120]: handleSpeedTopic() and handleAngleTopic() are *not* tested in this method, right?

**Review by iachella — COMMENTED** [review]
The code for unit testing looks good! I just left a question for a clarification. 

This little guy is an empty file, maybe we don't need it
external/googletest-release-1.10.0/googlemock/scripts/generator/cpp/__init__.py

Good work!

File: arduino/wrappers/Utils.h (hunk 1 above)
Comment (bassamEldesouki) [reply arduino/wrappers/Utils.h:120]: yeah, this is a typo .. it should be "not tested" good catch :)

File: external/googletest-release-1.10.0/CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
+# How to become a contributor and submit your own code
```
Comment (juliavankirk) [inline external/googletest-release-1.10.0/CONTRIBUTING.md:1]: Should these, along with some others,be gitignored?

File: external/googletest-release-1.10.0/LICENSE [hunk 3]
```diff
@@ -0,0 +1,28 @@
+Copyright 2008, Google Inc.
```
Comment (juliavankirk) [inline external/googletest-release-1.10.0/LICENSE:1]: gitignore

**Review by juliavankirk — COMMENTED** [review]
I

File: external/googletest-release-1.10.0/LICENSE (hunk 3 above)
Comment (platisd) [reply external/googletest-release-1.10.0/LICENSE:1]: Did you attend Imed's lecture last Monday? 😁
You probably shouldn't touch license files or snippets. 😅

File: external/googletest-release-1.10.0/CONTRIBUTING.md (hunk 2 above)
Comment (bassamEldesouki) [reply external/googletest-release-1.10.0/CONTRIBUTING.md:1]: I mean it will be hectic and will need much to go through 257 files in an external library and remove all these comments!

File: external/googletest-release-1.10.0/LICENSE (hunk 3 above)
Comment (juliavankirk) [reply external/googletest-release-1.10.0/LICENSE:1]: @platisd 
I don't think it's been touched per se, it seems like everything has just been committed unnecessarily.

File: external/googletest-release-1.10.0/CONTRIBUTING.md (hunk 2 above)
Comment (alrapal) [reply external/googletest-release-1.10.0/CONTRIBUTING.md:1]: @platisd 
Just out of curiosity, 
Is there a better way to include that library into the project (in the same way you can use gradle or maven for instance) or does it need to be "manually" imported? 
Because I don't see the smartCar library being added manually for this project, and as @bassamEldesouki mentioned, it is hectic to go though the whole library and chose what you want from it.

Comment (platisd) [reply external/googletest-release-1.10.0/CONTRIBUTING.md:1]: @alrapal there are different ways if you don't want to commit those external libraries, however, they are all more difficult than just including it.
One is to have CMake fetch it which is _sometimes_ problematic on Windows. You also need some extra CMake code for that.
Another is to have it as a `git submodule` which means that it would have to be fetched "manually" when cloning the repository.

Comment (juliavankirk) [reply external/googletest-release-1.10.0/CONTRIBUTING.md:1]: @bassamEldesouki here you go:

https://chromium.googlesource.com/external/github.com/google/googletest/+/refs/heads/master/.gitignore

File: external/googletest-release-1.10.0/LICENSE (hunk 3 above)
Comment (platisd) [reply external/googletest-release-1.10.0/LICENSE:1]: The file you suggest removing states the following:

> Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions aremet:
> * Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.

It is literally illegal to redistribute `googletest` code without the particular file/copyright notice.

File: external/googletest-release-1.10.0/CONTRIBUTING.md (hunk 2 above)
Comment (platisd) [reply external/googletest-release-1.10.0/CONTRIBUTING.md:1]: As for the Smartcar library, to use it one has to download it through the Arduino IDE or `arduino-cli`. In other words there's an extra "manual" step involved too.

Comment (alrapal) [reply external/googletest-release-1.10.0/CONTRIBUTING.md:1]: Thanks for these clarifications !

Comment (bassamEldesouki) [reply external/googletest-release-1.10.0/CONTRIBUTING.md:1]: @juliavankirk I believe as @platisd said that we shouldn't touch the external library. Also as I said it will be hectic to go through all these files. If we change one thing it might require other implementation and other steps will be involved and I don't know if we have enough time for that. One more important point is that this course is not about CMAKE and what you're asking is involving CMAKE deep understanding. I think this is not required.

Comment (bassamEldesouki) [reply external/googletest-release-1.10.0/CONTRIBUTING.md:1]: Remember, the point of this is just to make some unit tests for our code :)

File: external/googletest-release-1.10.0/LICENSE (hunk 3 above)
Comment (juliavankirk) [reply external/googletest-release-1.10.0/LICENSE:1]: @platisd 
honestly just landed my comment on there, since we very quickly glanced over this PR over Zoom
Obviously leaving the license alone, would you recommend we still leave everything else, mock tests included, in the commit? Should none of the committed files be gitignored?

(EDIT) But also to clarify:
You are right that we can do it this way. My point is why would we when instructions on how to get the code would suffice and be more clear for other developers and easier to maintain.

I believe it is wrong that we are required by the license agreement to include their code into our repo (if we make people download it separately then we are not distributing it). What they are referring to is the code for Google test itself, not our code that utilizes the tool. We wouldn’t be distributing Google test code in our product (if we gitignored/ uncommitted these files), we would simply be using it to verify and the verification related code is never distributed to the customer.

File: arduino/smartcar/smartcar.ino [hunk 4]
```diff
@@ -335,32 +338,34 @@ void connectivityLoss(){
...
+        case 2:
+            return (frontIRSensor.getDistance() == NO_OBSTACLE_VALUE);
+        case 3:
+            return (backIRSensor.getDistance() == NO_OBSTACLE_VALUE);
+        case 4:
+            return (rightIRSensor.getDistance() == NO_OBSTACLE_VALUE);
+        case 5:
+            return (leftIRSensor.getDistance() == NO_OBSTACLE_VALUE);
+        default:
+            return false;
```
Comment (alrapal) [inline arduino/smartcar/smartcar.ino:359]: Nice having changed this for a switch. I remember switch were not working with strings in c++ but here you are having numbers so that's good !

File: arduino/wrappers/Utils.cpp [hunk 5]
```diff
@@ -0,0 +1,153 @@
+/* This is the implementation of wrapper classes in header.
```
Comment (alrapal) [inline arduino/wrappers/Utils.cpp:1]: Is wrapper like a mock ? Or am I misunderstanding something?

File: external/googletest-release-1.10.0/LICENSE (hunk 3 above)
Comment (platisd) [reply external/googletest-release-1.10.0/LICENSE:1]: @juliavankirk taking the circumstances into consideration, I'd say that commiting the `googletest` sources as they currently are in this pull request makes sense.

File: arduino/wrappers/Utils.h [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+void MQTTMessageInput(MqttWrapper &mqtt, SerialWrapper &serial){
```
Comment (alrapal) [inline arduino/wrappers/Utils.h:121]: I am not familiar with testing in c++, but is it like in android or java where you have assertions?
Because I could not really see assertions in the tests case, so I am wondering how the test are considered passed or failed

File: external/googletest-release-1.10.0/LICENSE (hunk 3 above)
Comment (juliavankirk) [reply external/googletest-release-1.10.0/LICENSE:1]: @platisd by circumstances do you mean time constraints? I'd just like to know what best practices would be going forward, since I'm still learning as I go :)

**Review by alrapal — COMMENTED** [review]
I ran the tests locally and they all passed. 
Just a few questions below for my understanding of everything works !
Good job guys, I can imagine the research behind to wrap everything's up ! 🚀

File: external/googletest-release-1.10.0/LICENSE (hunk 3 above)
Comment (bassamEldesouki) [reply external/googletest-release-1.10.0/LICENSE:1]: @platisd It would be great if you please give us a feedback and your opinion what to keep in mind in the future unit testing :)

Comment (platisd) [reply external/googletest-release-1.10.0/LICENSE:1]: @juliavankirk the circumstances are the team's unfamiliarity with Cmake and therefore time.

In practice, even if the team was more experienced in Cmake (it's honestly not that hard btw but surely more complicated), the determining factor would be how easy it is for whatever build configuration system you are using to reliably fetch such external dependencies.
In professional projects I have seen such external dependencies fetched via Cmake (or equivalents), checked in like this pull request suggests and fetched as a git submodule.
In other words, all of these practices are potentially acceptable.

Comment (juliavankirk) [reply external/googletest-release-1.10.0/LICENSE:1]: @platisd I see, it's certainly the path of least resistance for us then. 
I'll keep your comments in mind going forward, thanks so much for explaining!

**Review by juliavankirk — APPROVED** [review]
Looks good and up to standard given @platisd comments

File: external/googletest-release-1.10.0/CONTRIBUTING.md (hunk 2 above)
Comment (ediz-genc) [reply external/googletest-release-1.10.0/CONTRIBUTING.md:1]: I thought we covered some important concerns about the unit testing today in the meeting for around 40 min. On another hand, I agree with the point that a deep understanding of the good libraries, licenses, and general coverage of unit testing is going to take place I believe in the next semester as a whole course. But as I said earlier in the meeting we are still happy to answer your further concerns as much as we can. 

Since we have some unit test now, it is also good clone the repository and play around with it :)

File: arduino/wrappers/Utils.cpp (hunk 5 above)
Comment (ediz-genc) [reply arduino/wrappers/Utils.cpp:1]: We used wrappers to be able to call the additional functions in related libiraires for the Smartcar, Arduino, and so on. Google Mock is related to dependencies.  I believe the link below will help you further :)

**Google Mock:**
https://chromium.googlesource.com/external/github.com/google/googletest/+/refs/tags/release-1.8.0/googlemock/docs/ForDummies.md

File: arduino/wrappers/Utils.h (hunk 6 above)
Comment (ediz-genc) [reply arduino/wrappers/Utils.h:121]: Yeah, they are similar. In this google unit test, we used, i.ex. EXPECT_EQ() this is a google function that expects two values in the parameters. These values should be equal. If the method parameter values which are given are equal then the test passed, otherwise in the case of inequality test fails.

**Comment by ediz-genc** [comment]
Does anybody else have some other concerns or anything you wonder? :)

**Review by iachella — APPROVED** [review]
Thanks for the clarifications in today's meeting, go ahead from me :)

**Review by alrapal — APPROVED** [review]
Good job! :D

File: external/googletest-release-1.10.0/LICENSE (hunk 3 above)
Comment (juliavankirk) [reply external/googletest-release-1.10.0/LICENSE:1]: @platisd 
hi again and sorry to be so annoying!  I promise to drop this line of questioning if you disapprove after this :)

Could we just delete all commits of googletest? Checking it all in is kind of messing with our stuff. Please see attached

Screen Shot 2021-05-27 at 1 04 48 PM

Proposed solution:
Provide instructions on how to get the code should suffice, and be more clear for other developers and easier to maintain. This way we can have someone download the repo from the source when they need it instead of keeping the tool in our repo. Maybe even integrate it into our Cmake build to do the download step automatically.

Comment (platisd) [reply external/googletest-release-1.10.0/LICENSE:1]: Not sure what was crashing but as explained all three solutions are generally accepted.

**Review by platisd — COMMENTED** [review]
Actually the `cmake-build-debug` directory shouldn't have  been version controlled so make sure you remove it and then gitignore it

File: external/googletest-release-1.10.0/LICENSE (hunk 3 above)
Comment (alrapal) [reply external/googletest-release-1.10.0/LICENSE:1]: (My browser was crashing when I tried to make the review directly. Had to wait a few minutes to let it load everything before being able to click on the button after crashing, but waiting was the solution in this case :p)

**Comment by ediz-genc** [comment]
> Actually the `cmake-build-debug` directory shouldn't have been version controlled

Strange... We deleted the directory before the commit as Ruthger mentioned about it once. But we will do it again. :) Thank you for pointing it out.

**Comment by ediz-genc** [comment]
> Actually the `cmake-build-debug` directory shouldn't have been version controlled so make sure you remove it and then gitignore it

@platisd , cmake-build-debug has been removed and gitignored. :)
